File: macos/Turkify/Turkify/EngineClient.swift
import Foundation

/// Python düzeltme motoruyla (`turkify serve --stdio`) köprü.
///
/// Motoru çocuk süreç olarak başlatır (sıcak tutar) ve satır-bazlı JSON
/// protokolüyle konuşur (bkz. ADR 0004):
///
///     istek :  {"id": 1, "text": "..."}
///     yanıt :  {"id": 1, "corrected": "..."} | {"id": 1, "error": "..."}
///
/// Ayarlar, motoru başlatırken **CLI bayrakları** olarak geçirilir (config.json
/// kullanılmaz — bkz. ADR 0007). Ayar değişince `restart(settings:)` ile yeni
/// bayraklarla yeniden başlatılır.
final class EngineClient {
    enum EngineError: Error {
        case engine(String)
        case badResponse
        case notRunning
    }

    private var process: Process?
    private var stdinHandle: FileHandle?

    private let queue = DispatchQueue(label: "com.turkify.engine")
    private var buffer = Data()
    private var pending: [Int: CheckedContinuation<String, Error>] = [:]
    private var nextID = 1

    var isRunning: Bool { process?.isRunning ?? false }

    /// Motoru verilen ayarlarla başlatır (önce çalışan varsa durdurur).
    func start(settings: AppSettings) throws {
        stop()
        let proc = Process()
        let launch = AppSettings.engineExecutable()
        proc.executableURL = URL(fileURLWithPath: launch.executable)
        proc.arguments = launch.prefixArgs + settings.serveArguments()
        Log.info("motor komutu: \(launch.executable) \((launch.prefixArgs + settings.serveArguments()).joined(separator: " "))")
        proc.terminationHandler = { p in
            Log.info("motor sonlandi (kod \(p.terminationStatus))")
        }

        let stdin = Pipe()
        let stdout = Pipe()
        let stderr = Pipe()
        proc.standardInput = stdin
        proc.standardOutput = stdout
        proc.standardError = stderr  // motor tanı mesajlarını yakala → Log sekmesi

        stdout.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            guard !data.isEmpty else { return }
            self?.queue.async { self?.ingest(data) }
        }
        stderr.fileHandleForReading.readabilityHandler = { handle in
            let data = handle.availableData
            guard !data.isEmpty, let text = String(data: data, encoding: .utf8) else { return }
            for line in text.split(whereSeparator: \.isNewline) where !line.isEmpty {
                Log.engine(String(line))
            }
        }

        try proc.run()
        process = proc
        stdinHandle = stdin.fileHandleForWriting
    }

    func restart(settings: AppSettings) throws {
        try start(settings: settings)
    }

    func stop() {
        if let proc = process, proc.isRunning { proc.terminate() }
        process = nil
        stdinHandle = nil
        queue.async {
            let waiting = self.pending
            self.pending.removeAll()
            self.buffer.removeAll()
            for (_, continuation) in waiting {
                continuation.resume(throwing: EngineError.notRunning)
            }
        }
    }

    /// Metni motora gönderir ve düzeltilmiş halini döndürür. Task iptal edilirse
    /// bekleyen istek bırakılır (motor isteği arka planda bitirir, yanıtı yok sayılır).
    func correct(_ text: String) async throws -> String {
        // id'yi önceden üret ki iptal işleyicisi doğru isteği temizleyebilsin.
        let id = queue.sync { () -> Int in
            let current = nextID
            nextID += 1
            return current
        }
        return try await withTaskCancellationHandler {
            try await withCheckedThrowingContinuation { continuation in
                queue.async {
                    guard let handle = self.stdinHandle, self.isRunning else {
                        continuation.resume(throwing: EngineError.notRunning)
                        return
                    }
                    self.pending[id] = continuation
                    let payload: [String: Any] = ["id": id, "text": text]
                    guard var line = try? JSONSerialization.data(withJSONObject: payload) else {
                        self.pending.removeValue(forKey: id)
                        continuation.resume(throwing: EngineError.badResponse)
                        return
                    }
                    line.append(0x0A)  // satır sonu
                    handle.write(line)
                }
            }
        } onCancel: {
            queue.async {
                if let continuation = self.pending.removeValue(forKey: id) {
                    continuation.resume(throwing: CancellationError())
                }
            }
        }
    }

    // MARK: - Gelen veriyi satırlara böl (queue üzerinde çalışır)

    private func ingest(_ data: Data) {
        buffer.append(data)
        while let newline = buffer.firstIndex(of: 0x0A) {
            let lineData = buffer.subdata(in: buffer.startIndex..<newline)
            buffer.removeSubrange(buffer.startIndex...newline)
            handleLine(lineData)
        }
    }

    private func handleLine(_ data: Data) {
        guard
            let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            let id = obj["id"] as? Int,
            let continuation = pending.removeValue(forKey: id)
        else { return }

        if let corrected = obj["corrected"] as? String {
            continuation.resume(returning: corrected)
        } else if let error = obj["error"] as? String {
            continuation.resume(throwing: EngineError.engine(error))
        } else {
            continuation.resume(throwing: EngineError.badResponse)
        }
    }
}

File: windows/Turkify/EngineClient.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Text.Json;

namespace Turkify;

/// Motor hata türleri.
public sealed class EngineException(string message) : Exception(message);

public sealed class EngineNotRunningException() : Exception("Motor çalışmıyor");

public sealed class EngineBadResponseException() : Exception("Motor yanıtı çözülemedi");

/// Python düzeltme motoruyla (<c>turkify serve --stdio</c>) köprü.
///
/// Motoru çocuk süreç olarak başlatır (sıcak tutar) ve satır-bazlı JSON
/// protokolüyle konuşur (bkz. ADR 0004):
///
///     istek :  {"id": 1, "text": "..."}
///     yanıt :  {"id": 1, "corrected": "..."} | {"id": 1, "error": "..."}
///
/// Ayarlar, motoru başlatırken **CLI bayrakları** olarak geçirilir (config.json
/// kullanılmaz — bkz. ADR 0007). Ayar değişince <see cref="Restart"/> ile yeni
/// bayraklarla yeniden başlatılır. (macOS EngineClient.swift'in karşılığı.)
public sealed class EngineClient : IDisposable
{
    private static readonly UTF8Encoding Utf8NoBom = new(encoderShouldEmitUTF8Identifier: false);

    private readonly object _gate = new();
    private readonly ConcurrentDictionary<int, TaskCompletionSource<string>> _pending = new();
    private Process? _process;
    private int _nextId = 1;

    public bool IsRunning
    {
        get
        {
            lock (_gate)
            {
                return _process is { HasExited: false };
            }
        }
    }

    /// Motoru verilen ayarlarla başlatır (önce çalışan varsa durdurur).
    public void Start(AppSettings settings)
    {
        Stop();

        (string executable, IReadOnlyList<string> prefixArgs) = AppSettings.EngineExecutable();
        var startInfo = new ProcessStartInfo
        {
            FileName = executable,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
            StandardOutputEncoding = Utf8NoBom,
            StandardErrorEncoding = Utf8NoBom,
            StandardInputEncoding = Utf8NoBom,
        };
        foreach (string arg in prefixArgs)
        {
            startInfo.ArgumentList.Add(arg);
        }

        foreach (string arg in settings.ServeArguments())
        {
            startInfo.ArgumentList.Add(arg);
        }

        Log.Info($"motor komutu: {executable} {string.Join(' ', startInfo.ArgumentList)}");

        var process = new Process { StartInfo = startInfo, EnableRaisingEvents = true };
        process.OutputDataReceived += OnOutput;
        process.ErrorDataReceived += OnError;
        process.Exited += (_, _) => Log.Info("motor sonlandi");

        process.Start();
        process.BeginOutputReadLine();
        process.BeginErrorReadLine();

        lock (_gate)
        {
            _process = process;
        }
    }

    public void Restart(AppSettings settings) => Start(settings);

    /// Motora ateşle-unut bir <c>{"cmd":"reload"}</c> gönderir; motor ayarları ve
    /// korumalı kelime önbelleğini tazeler (model sıcak kalır). Yanıt id taşımadığından
    /// yok sayılır. Motor çalışmıyorsa sessizce atlanır.
    public void Reload()
    {
        WriteLine("{\"cmd\":\"reload\"}");
    }

    public void Stop()
    {
        Process? process;
        lock (_gate)
        {
            process = _process;
            _process = null;
        }

        if (process is not null)
        {
            try
            {
                if (!process.HasExited)
                {
                    // stdin EOF → motor temiz çıkar (serve_stdio EOF'ta döner).
                    process.StandardInput.Close();
                    if (!process.WaitForExit(1000))
                    {
                        process.Kill();
                    }
                }
            }
            catch (InvalidOperationException)
            {
                // süreç zaten gitmiş
            }
            finally
            {
                process.Dispose();
            }
        }

        // Bekleyen tüm istekleri "çalışmıyor" ile sonlandır.
        foreach (KeyValuePair<int, TaskCompletionSource<string>> entry in _pending)
        {
            if (_pending.TryRemove(entry.Key, out TaskCompletionSource<string>? tcs))
            {
                tcs.TrySetException(new EngineNotRunningException());
            }
        }
    }

    /// Metni motora gönderir ve düzeltilmiş halini döndürür. İptal edilirse bekleyen
    /// istek bırakılır (motor isteği arka planda bitirir, yanıtı yok sayılır).
    public async Task<string> CorrectAsync(string text, CancellationToken cancellationToken = default)
    {
        if (!IsRunning)
        {
            throw new EngineNotRunningException();
        }

        int id = Interlocked.Increment(ref _nextId);
        var tcs = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
        _pending[id] = tcs;

        string payload = JsonSerializer.Serialize(new RequestPayload(id, text));
        if (!WriteLine(payload))
        {
            _pending.TryRemove(id, out _);
            throw new EngineNotRunningException();
        }

        await using (cancellationToken.Register(() =>
        {
            if (_pending.TryRemove(id, out TaskCompletionSource<string>? pendingTcs))
            {
                pendingTcs.TrySetCanceled(cancellationToken);
            }
        }))
        {
            return await tcs.Task.ConfigureAwait(false);
        }
    }

    private bool WriteLine(string line)
    {
        lock (_gate)
        {
            if (_process is null || _process.HasExited)
            {
                return false;
            }

            try
            {
                // Motor satırı strip() eder; "\n" yeterli (Windows "\r\n" de sorun değil).
                _process.StandardInput.Write(line);
                _process.StandardInput.Write('\n');
                _process.StandardInput.Flush();
                return true;
            }
            catch (IOException)
            {
                return false;
            }
            catch (ObjectDisposedException)
            {
                return false;
            }
        }
    }

    private void OnOutput(object sender, DataReceivedEventArgs e)
    {
        if (string.IsNullOrEmpty(e.Data))
        {
            return;
        }

        HandleLine(e.Data);
    }

    private void OnError(object sender, DataReceivedEventArgs e)
    {
        if (string.IsNullOrEmpty(e.Data))
        {
            return;
        }

        Log.Engine(Log.StripLeadingTimestamp(e.Data));
    }

    private void HandleLine(string line)
    {
        int? id;
        string? corrected;
        string? error;
        try
        {
            using JsonDocument document = JsonDocument.Parse(line);
            JsonElement root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object || !root.TryGetProperty("id", out JsonElement idElement))
            {
                return; // id taşımayan yanıt (ör. reload {"ok":true}) yok sayılır
            }

            id = idElement.GetInt32();
            corrected = root.TryGetProperty("corrected", out JsonElement c) ? c.GetString() : null;
            error = root.TryGetProperty("error", out JsonElement err) ? err.GetString() : null;
        }
        catch (JsonException)
        {
            return; // bozuk satır yok sayılır
        }

        if (id is null || !_pending.TryRemove(id.Value, out TaskCompletionSource<string>? tcs))
        {
            return;
        }

        if (corrected is not null)
        {
            tcs.TrySetResult(corrected);
        }
        else if (error is not null)
        {
            tcs.TrySetException(new EngineException(error));
        }
        else
        {
            tcs.TrySetException(new EngineBadResponseException());
        }
    }

    public void Dispose() => Stop();

    private sealed record RequestPayload(int id, string text);
}
